src/runtime/type.ts

```diff
@@ -381,42 +381,42 @@ export class TypeVariable extends Type {
 }
 
 
-export class CollectionType extends Type {
-  readonly collectionType: Type;
+export class CollectionType<T extends Type> extends Type {
+  readonly collectionType: T;
 
-  constructor(collectionType: Type) {
+  constructor(collectionType: T) {
     super('Collection');
     this.collectionType = collectionType;
   }
 
-  get isCollection() {
+  get isCollection(): boolean {
     return true;
   }
 
-  mergeTypeVariablesByName(variableMap: Map<string, Type>) {
+  mergeTypeVariablesByName(variableMap: Map<string, Type>): CollectionType<Type> {
```

Comment:
CollectionType<T>?

Reply:
In this case, `const result` is of type `Type`, so `result.collectionOf()` has type `Collection<Type>`.

So the best we can do here is to constrain the return type to `CollectionType<Type>`.